Revert TextInput dictation `\uFFFC` to `_UITextPlaceholderAttachment`

hellohublot opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

When we start using the dictation on the keyboard, the system will automatically generate a `_UITextPlaceholderAttachment`, when the speech recognition ends, the system will delete `_UITextPlaceholderAttachment` and replace it with the recognized text, but when we pass `_UITextPlaceholderAttachment` to Javascript, we will force convert `_UITextPlaceholderAttachment` object to `\uFFFC`, so after the system speech recognition is completed, the `_UITextPlaceholderAttachment` cannot be found.

## Changelog:

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

[IOS][FIXED] - Reverse TextInput `\uFFFC` to private class `_UITextPlaceholderAttachment` object

## Test Plan:

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

Test Code
```javascript
<TextInput
    ...
    onChangeText={(text) => this.setState({ text }, () => console.log(text.length))}
    value={this.state.text}
/>
```

Case 1: 
1. Start voice input, but don't speak, 
2. Verfiry the final output of the console is 0

Case 2: 
1. Start voice input, say a word randomly
2. Verify that the final output of the console is consistent with the length of the word

Case 3: 
1. Start voice input, use voice to input sentences, and click the keyboard to input text at the same time
2. Verify that the final output of the console is consistent with the length of the sentence

dmytrorykun commented:
Hi @hellohublot, thank you for the PR. Unfortunately we can't merge it because it is using private API `_UITextPlaceholderAttachment`.
- Could you please add more detailed description of the problem? What you are doing, the expected behaviour, the actual behaviour.
- As for the proper fix, firstly I would try to understand why the following happens 👇
> the system will delete _UITextPlaceholderAttachment and replace it with the recognized text, but when we pass _UITextPlaceholderAttachment to Javascript, we will force convert _UITextPlaceholderAttachment object to \uFFFC

hellohublot commented:
@dmytrorykun 

Thanks, everything has been changed, could you please review it again

dmytrorykun commented:
@hellohublot could you please rebase this PR on the latest `master`.

hellohublot commented:
@dmytrorykun Thanks, done.

hellohublot commented:
@dmytrorykun Hello, Is there anything I can help you with? Please feel free to ask

dmytrorykun reviewed (changes requested):
Hey @hellohublot. `typingAttributesBeforeInsertion` is still private API. Even though it is not prefixed with an underscore, it is not publicly documented. Developers are not supposed to use private APIs because it may change or stop working any time without warning.
Try using this routine to detect dictation. Does it work for all the cases?
```
  BOOL isDictationActive = NO;
  for (int characterIndex = 0; characterIndex < oldText.string.length; characterIndex++) {
    if ([oldText.string characterAtIndex:characterIndex] == NSAttachmentCharacter) {
      isDictationActive = YES;
    }
  }
```

hellohublot commented:
@dmytrorykun  Thanks for your reply, but I guess that won't work, because we only want to judge whether there is a `_UITextPlaceholderAttachment`, so when other `NSTextAttachment` exists, we can't set `isDictatingRunning` to `true`, and there are two `[NSTextAttachment new]` in our code

1. Since it doesn't start with an `underscore`, I think this can pass the machine review, because the class here is `id`, so Apple can't judge whether `typingAttributesBeforeInsertion` is a method written by ourselves

2. `NSOriginalFont` in our code can pass the `machine review`, so `typingAttributesBeforeInsertion` can also pass the machine review

3. I tried the following two methods, but they can't be used to judge
https://developer.apple.com/documentation/uikit/uitextinput/1614466-insertdictationresultplaceholder
https://developer.apple.com/documentation/uikit/uitextinput/1614493-frame

4. Apple only opened `dictationRecordingDidEnd` and `dictationRecognitionFailed`, but didn't open `dictationRecordingDidBegin`, so I think `typingAttributesBeforeInsertion` is the best way

5. `typingAttributesBeforeInsertion` is an attribute that has been supported since `iOS7+`, so it is stable enough. If it is replaced by other methods after many version updates, we can continue to modify it to other methods, because we really can't find a public api

jpdriver commented:
could we move `isDictationRunning` to the View(s) and set it to `YES` in `dictationRecordingDidEnd`?

it seems like since Apple apparently only wants us to know when dictation ended or failed, that could be the way to go? 🤔

hellohublot commented:
@jpdriver 

Thank you for your suggestion, but I guess that _UITextPlaceholderAttachment corresponds to the circle in loading, so dictationRecordingDidEnd has not been executed at this time, because when _UITextPlaceholderAttachm is displayed, voice recognition is in progress

https://github.com/facebook/react-native/assets/20135674/00c72acd-9d1d-479d-8688-eff08e21dd0e

hellohublot commented:
@jpdriver Thanks for your suggestion, I mistook `dictationRecordingDidEnd` for `dictationRecognitionDidEnd` so I ignored it, I'll try it tomorrow if it works with `insertDictationResult` or `removeDictationResultPlaceholder`

hellohublot commented:
@dmytrorykun @jpdriver  Thanks for your suggestions, I have changed it to `dictationRecordingDidEnd`, I thought it was `dictationRecognitionDidEnd` before, because it is too similar to `dictationRecognitionFailed`

I also modified the PR description, could you please help me to review it again

hellohublot commented:
@dmytrorykun Hi, Is there anything I can help you with? Please feel free to ask
